File: packages/less/lib/less/tree/variable.js
// @ts-check
/** @import { EvalContext, FileInfo } from './node.js' */
import Node from './node.js';
import Call from './call.js';
import Ruleset from './ruleset.js';

class Variable extends Node {
    get type() { return 'Variable'; }

    /**
     * @param {string} name
     * @param {number} [index]
     * @param {FileInfo} [currentFileInfo]
     */
    constructor(name, index, currentFileInfo) {
        super();
        this.name = name;
        this._index = index;
        this._fileInfo = currentFileInfo;
        /** @type {boolean | undefined} */
        this.evaluating = undefined;
    }

    /**
     * @param {EvalContext} context
     * @returns {Node}
     */
    eval(context) {
        let variable, name = this.name;

        if (name.indexOf('@@') === 0) {
            name = `@${new Variable(name.slice(1), this.getIndex(), this.fileInfo()).eval(context).value}`;
        }

        if (this.evaluating) {
            throw { type: 'Name',
                message: `Recursive variable definition for ${name}`,
                filename: this.fileInfo().filename,
                index: this.getIndex() };
        }

        this.evaluating = true;

        variable = this.find(context.frames, function (frame) {
            const v = /** @type {Ruleset} */ (frame).variable(name);
            if (v) {
                if (v.important) {
                    const importantScope = context.importantScope[context.importantScope.length - 1];
                    importantScope.important = v.important;
                }
                // If in calc, wrap vars in a function call to cascade evaluate args first
                if (context.inCalc) {
                    return (new Call('_SELF', [v.value], 0, undefined)).eval(context);
                }
                else {
                    return v.value.eval(context);
                }
            }
        });
        if (variable) {
            this.evaluating = false;
            return variable;
        } else {
            throw { type: 'Name',
                message: `variable ${name} is undefined`,
                filename: this.fileInfo().filename,
                index: this.getIndex() };
        }
    }

    /**
     * @param {Node[]} obj
     * @param {(frame: Node) => Node | undefined} fun
     * @returns {Node | null}
     */
    find(obj, fun) {
        for (let i = 0, r; i < obj.length; i++) {
            r = fun.call(obj, obj[i]);
            if (r) { return r; }
        }
        return null;
    }
}

export default Variable;

File: packages/less/src/less/go_parser/variable.go
package go_parser

import "fmt"

// Frame represents a scope frame that can look up variables
type Frame interface {
	Variable(name string) map[string]any
}

// Variable represents a variable node in the Less AST
type Variable struct {
	*Node
	name      string
	_index    int
	_fileInfo map[string]any
	evaluating bool
}

// NewVariable creates a new Variable instance
func NewVariable(name string, index int, currentFileInfo map[string]any) *Variable {
	return &Variable{
		Node:      NewNode(),
		name:      name,
		_index:    index,
		_fileInfo: currentFileInfo,
	}
}

// GetType returns the node type
func (v *Variable) GetType() string {
	return "Variable"
}

// GetIndex returns the node's index
func (v *Variable) GetIndex() int {
	return v._index
}

// FileInfo returns the node's file information
func (v *Variable) FileInfo() map[string]any {
	return v._fileInfo
}

// GetName returns the variable name
func (v *Variable) GetName() string {
	return v.name
}

// Eval evaluates the variable
func (v *Variable) Eval(context EvalContext) (any, error) {
	name := v.name

	if len(name) >= 2 && name[:2] == "@@" {
		innerVar := NewVariable(name[1:], v.GetIndex(), v.FileInfo())
		innerResult, err := innerVar.Eval(context)
		if err != nil {
			return nil, err
		}
		
		// Convert innerResult to string and prepend "@"
		name = "@" + fmt.Sprintf("%v", innerResult.(map[string]any)["value"])
	}

	if v.evaluating {
		return nil, fmt.Errorf("name: recursive variable definition for %s (index: %d, filename: %s)", 
			name, v.GetIndex(), v.FileInfo()["filename"])
	}

	v.evaluating = true

	// Get frames from context
	frames, ok := context.(interface{ GetFrames() []Frame })
	if !ok {
		return nil, fmt.Errorf("context does not implement GetFrames")
	}

	// Find variable in frames
	for _, frame := range frames.GetFrames() {
		if varResult := frame.Variable(name); varResult != nil {
			// Handle important flag if present
			if importantVal, ok := varResult["important"].(bool); ok && importantVal {
				importantScopes, ok := context.(interface{ GetImportantScope() []map[string]bool })
				if ok && len(importantScopes.GetImportantScope()) > 0 {
					lastScope := importantScopes.GetImportantScope()[len(importantScopes.GetImportantScope())-1]
					lastScope["important"] = true
				}
			}

			// Get value from result
			val, ok := varResult["value"]
			if !ok {
				continue
			}

			// If in calc, wrap vars in a function call to cascade evaluate args first
			if isInCalc, ok := context.(interface{ IsInCalc() bool }); ok && isInCalc.IsInCalc() {
				selfCall := NewCall("_SELF", []any{val}, v.GetIndex(), v.FileInfo())
				v.evaluating = false
				return selfCall, nil
			}

			// Evaluate value
			if evalable, ok := val.(interface{ Eval(EvalContext) (any, error) }); ok {
				v.evaluating = false
				return evalable.Eval(context)
			}
		}
	}

	v.evaluating = false
	return nil, fmt.Errorf("name: variable %s is undefined (index: %d, filename: %s)", 
		name, v.GetIndex(), v.FileInfo()["filename"])
} 